frame/uniswap-v2/src/lib.rs

```diff
@@ -351,6 +390,108 @@ pub mod pallet {
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			let current_timestamp = T::Time::now();
+			Pools::<T>::iter()
+				.filter(|(pool_id, pool)| T::EnabledPoolTWAP::contains(&(pool.pair, *pool_id)))
+				.fold(Weight::zero(), |weight, (pool_id, _)| {
+					let rate_base = Self::do_get_exchange_rate(pool_id, false);
+					let rate_quote = Self::do_get_exchange_rate(pool_id, true);
+					match (rate_base, rate_quote) {
+						(Ok(rate_base), Ok(rate_quote)) => PriceCumulativesState::<T>::try_mutate(
+							pool_id,
+							|price_cumulatives| -> Result<Weight, DispatchError> {
+								match *price_cumulatives {
+									// Initialization
+									None => {
+										let (price_cumulative_base, price_cumulative_quote) = (
+											compute_initial_price_cumulative::<T::Convert, _>(
+												rate_base,
+											)?,
+											compute_initial_price_cumulative::<T::Convert, _>(
+												rate_quote,
+											)?,
+										);
+										*price_cumulatives = Some(PriceCumulatives {
+											timestamp: current_timestamp,
+											price_cumulative_base,
+											price_cumulative_quote,
+										});
+										TWAPState::<T>::insert(
+											pool_id,
+											TWAP {
+												average_price_base: rate_base,
+												average_price_quote: rate_quote,
+											},
+										);
+										Ok(weight + 1)
+									},
+									// Update
+									Some(PriceCumulatives {
+										timestamp: previous_timestamp,
+										price_cumulative_base: previous_price_cumulative_base,
+										price_cumulative_quote: previous_price_cumulative_quote,
+									}) => {
+										let f = |previous_cumulative_price, exchange_rate| {
+											compute_next_price_cumulative::<T::Convert, _, _>(
+												previous_timestamp,
+												previous_cumulative_price,
+												current_timestamp,
+												exchange_rate,
+											)
+										};
+										let (
+											(elapsed, price_cumulative_base),
+											(_, price_cumulative_quote),
+										) = (
+											f(previous_price_cumulative_base, rate_base)?,
+											f(previous_price_cumulative_quote, rate_quote)?,
+										);
+										if elapsed >= T::TWAPInterval::get() {
+											*price_cumulatives = Some(PriceCumulatives {
+												timestamp: current_timestamp,
+												price_cumulative_base,
+												price_cumulative_quote,
+											});
+											let avg = |current: T::Balance,
+											           previous: T::Balance|
+											 -> Result<Rate, DispatchError> {
+												compute_price_average::<T::Convert, _, _>(
+													current, previous, elapsed,
+												)
+											};
+											let average_price_base = avg(
+												price_cumulative_base,
+												previous_price_cumulative_base,
+											)?;
+											let average_price_quote = avg(
+												price_cumulative_quote,
+												previous_price_cumulative_quote,
+											)?;
+											TWAPState::<T>::insert(
```

Comment:
remove from storage

Reply:
why? it is just updating the average_price

Reply:
it is current and actual price. it is not average.

https://en.wikipedia.org/w/index.php?title=Time-weighted_average_price&diff=1080367079&oldid=1033791488

https://datalore.jetbrains.com/notebook/jSkxTlpnHIMBV2lHlhBXLO/ONTTwh1dCTwxn7pvAePkFZ/

Reply:
I think it is current actual price only right, if you check the function.  May be naming is different.

Reply:
i think this price should be RPC, not state

Reply:
and not named average